pkg/webhook/bootstrap.go

```diff
@@ -0,0 +1,315 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/ghodss/yaml"
+
+	"k8s.io/api/admissionregistration/v1beta1"
+	admissionregistration "k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/cert"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/cert/writer"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/types"
+)
+
+// setDefault does defaulting for the Server.
+func (s *Server) setDefault() {
+	if len(s.Name) == 0 {
+		s.Name = "default-admission-server"
+	}
+	if s.registry == nil {
+		s.registry = map[string]Webhook{}
+	}
+	if s.sMux == nil {
+		s.sMux = http.DefaultServeMux
+	}
+	if s.Port <= 0 {
+		s.Port = 443
+	}
+	if len(s.CertDir) == 0 {
+		s.CertDir = path.Join("admission-server", "cert")
+	}
+	if s.Host == nil && s.Service == nil {
+		varString := "localhost"
+		s.Host = &varString
+	}
+	if s.CertProvisioner == nil {
+		s.CertProvisioner = &cert.Provisioner{Client: s.Client}
+	}
+}
+
+// bootstrap returns the configuration of admissionWebhookConfiguration in yaml format if dryrun is true.
+// Otherwise, it creates the the admissionWebhookConfiguration objects and service if any.
+// It also provisions the certificate for your admission server.
+func (s *Server) bootstrap(dryrun bool) ([]byte, error) {
+	var err error
+	s.once.Do(func() {
+		s.setDefault()
+		if s.Client != nil {
+			return
+		}
+		s.Client, err = client.New(config.GetConfigOrDie(), client.Options{})
```

Comment:
We generally avoid initializing `OrDie` things in the defaulter.  We should be passing this in instead.